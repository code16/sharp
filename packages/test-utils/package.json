{
    "name": "@sharp/test-utils",
    "private": true,
    "main": "src/index.js",
    "version": "1.0.0"
}