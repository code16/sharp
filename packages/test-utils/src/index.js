export { default as MockInjections } from './MockInjections';
export { default as MockI18n } from './MockI18n';
export { default as QueryComponent } from './QueryComponent';
export { default as MockTransitions } from './MockTransitions';
export { mockSFC, unmockSFC } from './mockSFC';
export { mockProperty, unmockProperty, setter } from './mock-utils';
export { nextRequestFulfilled } from './moxios-utils';

export function wait(delay) {
    return new Promise(resolve => {
        setTimeout(resolve, delay);
    });
}