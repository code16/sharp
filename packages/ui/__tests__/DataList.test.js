import merge from 'lodash/merge';

import SharpDataList from '../src/components/DataList.vue';
import { shallowMount } from '@vue/test-utils';

describe('DataList', ()=>{

    function createWrapper({ ...options }={}) {
        return shallowMount(SharpDataList, {
            // language=Vue
            stubs: {
                DataListRow:
                    `<div class="MOCKED_SharpDataListRow" v-bind="$props">
                        <template v-for="column in columns">
                            <slot name="cell" :column="column" />
                        </template>
                    </div>`,
                Draggable:
                    `<div class="MOCKED_Draggable">
                        <slot />
                    </div>`,
            },
            ...options,
        });
    }

    function setReorderActive(wrapper, reorderActive) {
        const watcherSpy = jest.spyOn(wrapper.vm, 'handleReorderActiveChanged').mockImplementation();
        wrapper.setProps({
            reorderActive
        });
        watcherSpy.mockRestore();
    }

    function withDefaultMocks(options={}) {
        return merge({
            computed: {
                isEmpty: ()=>false,
            }
        }, options);
    }

    test('mount empty', ()=>{
        const wrapper = createWrapper(withDefaultMocks({
            computed: {
                isEmpty: ()=>true,
            },
            slots: {
                empty: 'Empty text',
            },
        }));
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('mount items', ()=>{
        const wrapper = createWrapper(withDefaultMocks());
        wrapper.setProps({
            items: [{ id:1 }],
        });
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('mount header', ()=>{
        const wrapper = createWrapper(withDefaultMocks());
        const header = wrapper.find({ ref:'head' });
        const columns = [{ key:'col1', label:'col1' }, { key:'col2', label:'col2', sortable: true }];

        expect(header.html()).toMatchSnapshot('default');
        wrapper.setProps({
            columns: [...columns],
        });
        expect(header.html()).toMatchSnapshot('columns');
        wrapper.setProps({
            columns: [...columns],
            sort: 'col2',
        });
        expect(header.html()).toMatchSnapshot('sort selected');
        wrapper.setProps({
            columns: [...columns],
            sort: 'col2',
            dir: 'asc',
        });
        expect(header.html()).toMatchSnapshot('sort ascending');
    });

    test('mount items slot', ()=>{
        const wrapper = createWrapper(withDefaultMocks({
            scopedSlots: {
                item: `<div v-bind="props">{{ props.item.text }}</div>`,
            },
        }));
        wrapper.setProps({
            items: [{ id:1, text:'mockedText' }],
        });
        expect(wrapper.html()).toMatchSnapshot();
    });

    test('mount pagination', ()=>{
        const wrapper = createWrapper(withDefaultMocks({
            computed: {
                isEmpty: ()=>true,
                hasPagination: ()=>true
            },
        }));
        wrapper.setProps({
            totalCount: 10,
            pageSize: 3,
            page: 2,
        });
        expect(wrapper.html()).toMatchSnapshot();
    });

    describe('watch', ()=>{
        test('reorderActive', ()=>{
            const wrapper = createWrapper();
            wrapper.setMethods({
                handleReorderActiveChanged: jest.fn()
            });
            wrapper.setProps({
                reorderActive: true,
            });
            expect(wrapper.vm.handleReorderActiveChanged).toHaveBeenCalledWith(true);
        });
    });

    describe('computed', ()=>{
        test('hasPagination', ()=>{
            const wrapper = createWrapper();
            expect(wrapper.vm.hasPagination).toBe(false);
            wrapper.setProps({
                paginated: true,
            });
            expect(wrapper.vm.hasPagination).toBe(false);
            wrapper.setProps({
                paginated: true,
                totalCount: 10,
                pageSize: 3,
            });
            expect(wrapper.vm.hasPagination).toBe(true);
        });
        test('draggableOptions', ()=>{
            const wrapper = createWrapper();
            expect(wrapper.vm.draggableOptions).toEqual({
                disabled: true,
            });
            setReorderActive(wrapper, true);
            expect(wrapper.vm.draggableOptions).toEqual({
                disabled: false,
            });
        });
        test('currentItems', ()=>{
            const wrapper = createWrapper();
            wrapper.setProps({
                items: [{ label:'item' }],
            });
            expect(wrapper.vm.currentItems).toEqual([{ label:'item' }]);
            setReorderActive(wrapper, true);
            wrapper.setData({
                reorderedItems: [{ label: 'reorderedItem' }],
            });
            expect(wrapper.vm.currentItems).toEqual([{ label: 'reorderedItem' }]);
        });
        test('isEmpty', ()=>{
            const wrapper = createWrapper();
            expect(wrapper.vm.isEmpty).toBe(true);
            wrapper.setProps({
                items:[{ id:1 }],
            });
            expect(wrapper.vm.isEmpty).toBe(false);
        });
    });

    describe('methods', ()=>{
        test('handleItemsChanged', ()=>{
            const wrapper = createWrapper();
            wrapper.vm.handleItemsChanged('reorderedItems');
            expect(wrapper.vm.reorderedItems).toEqual('reorderedItems');
            expect(wrapper.emitted('change')[0]).toEqual(['reorderedItems']);
        });

        test('handleSortChanged', ()=>{
            let wrapper;
            wrapper = createWrapper();
            wrapper.vm.handleSortClicked('name');
            expect(wrapper.emitted('sort-change')[0]).toEqual([{ prop:'name', dir: 'asc' }]);

            wrapper = createWrapper();
            wrapper.setProps({
                sort: 'name',
                dir: 'asc',
            });
            wrapper.vm.handleSortClicked('name');
            expect(wrapper.emitted('sort-change')[0]).toEqual([{ prop:'name', dir: 'desc' }]);
        });

        test('handlePageChanged', ()=>{
            const wrapper = createWrapper();
            wrapper.vm.handlePageChanged(1);
            expect(wrapper.emitted('page-change')[0]).toEqual([1]);
        });

        test('handleReorderActiveChanged', ()=>{
            const wrapper = createWrapper();
            wrapper.setProps({
                items: [{ id:1 }]
            });
            wrapper.vm.handleReorderActiveChanged(true);
            expect(wrapper.vm.reorderedItems).toEqual([{ id:1 }]);
            wrapper.vm.handleReorderActiveChanged(false);
            expect(wrapper.vm.reorderedItems).toEqual(null);
        });
    });
});