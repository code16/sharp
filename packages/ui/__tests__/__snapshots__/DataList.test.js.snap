// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataList mount empty 1`] = `
<div class="SharpDataList">
	<div class="SharpDataList__empty p-3">Empty text</div>
	<!---->
</div>
`;

exports[`DataList mount header: columns 1`] = `
<div class="SharpDataList__thead">
	<div row="[object Object]" header="true" class="MOCKED_SharpDataListRow" columns="[object Object],[object Object]">
		<span>col1</span>
		<!---->
		<span>col2</span>
		<svg width="10" height="5" viewBox="0 0 10 5" fill-rule="evenodd" class="SharpDataList__caret">
			<path d="M10 0L5 5 0 0z"></path>
		</svg>
		<a href="" class="SharpDataList__sort-link"></a>
	</div>
</div>
`;

exports[`DataList mount header: default 1`] = `
<div class="SharpDataList__thead">
	<div row="[object Object]" header="true" class="MOCKED_SharpDataListRow"></div>
</div>
`;

exports[`DataList mount header: sort ascending 1`] = `
<div class="SharpDataList__thead">
	<div row="[object Object]" header="true" class="MOCKED_SharpDataListRow" columns="[object Object],[object Object]">
		<span>col1</span>
		<!---->
		<span>col2</span>
		<svg width="10" height="5" viewBox="0 0 10 5" fill-rule="evenodd" class="SharpDataList__caret SharpDataList__caret--selected SharpDataList__caret--ascending">
			<path d="M10 0L5 5 0 0z"></path>
		</svg>
		<a href="" class="SharpDataList__sort-link"></a>
	</div>
</div>
`;

exports[`DataList mount header: sort selected 1`] = `
<div class="SharpDataList__thead">
	<div row="[object Object]" header="true" class="MOCKED_SharpDataListRow" columns="[object Object],[object Object]">
		<span>col1</span>
		<!---->
		<span>col2</span>
		<svg width="10" height="5" viewBox="0 0 10 5" fill-rule="evenodd" class="SharpDataList__caret SharpDataList__caret--selected">
			<path d="M10 0L5 5 0 0z"></path>
		</svg>
		<a href="" class="SharpDataList__sort-link"></a>
	</div>
</div>
`;

exports[`DataList mount items 1`] = `
<div class="SharpDataList">
	<div class="SharpDataList__table SharpDataList__table--border">
		<div class="SharpDataList__thead">
			<div row="[object Object]" header="true" class="MOCKED_SharpDataListRow"></div>
		</div>
		<div class="SharpDataList__tbody">
			<div class="MOCKED_Draggable">
				<div row="[object Object]" class="MOCKED_SharpDataListRow"></div>
			</div>
		</div>
	</div>
	<!---->
</div>
`;

exports[`DataList mount items slot 1`] = `
<div class="SharpDataList">
	<div class="SharpDataList__table SharpDataList__table--border">
		<div class="SharpDataList__thead">
			<div row="[object Object]" header="true" class="MOCKED_SharpDataListRow"></div>
		</div>
		<div class="SharpDataList__tbody">
			<div class="MOCKED_Draggable">
				<div item="[object Object]">mockedText</div>
			</div>
		</div>
	</div>
	<!---->
</div>
`;

exports[`DataList mount pagination 1`] = `
<div class="SharpDataList">
	<div class="SharpDataList__empty p-3"></div>
	<div class="SharpDataList__pagination-container">
		<pagination-stub minpageendbuttons="3" limit="7" totalrows="10" perpage="3" value="2"></pagination-stub>
	</div>
</div>
`;
